### PR TITLE
migrate frontend to typescript (#60)

### PR DESCRIPTION
**Pull request by scottrowan**
moving to typescript:

major things,
- .js, .jsx files now .ts, .tsx
- tsconfig.json added to root dir
- we never pass "props" always a typed object with a defined interface
- interfaces defined and xported from the service files
- components changed to pass down typed objects rather than individual fields
- all function args and local variables / useStates etc have types defined

sorry the diff isn't so easy to read because its technically deleting / creating new files, not much I can do about that

all my type interfaces are named OpenAI*, such as OpenAIMessage, OpenAIResponse, OpenAIDefence etc, feel free to propose better names

### CODE REVIEW

**Comment by scottrowan** [comment]
noticed this will almost certain clash with George's PR, sure merge that first then I'll make a few changes here

File: frontend/src/components/ChatBox/ChatBox.tsx [hunk 1]
```diff
@@ -1,26 +1,36 @@
...
-function ChatBox(props) {
-  const [isSendingMessage, setIsSendingMessage] = useState(false);
-  const [messages, setMessages] = useState([]);
+function ChatBox(
+  this: any,
+  {
+    setEmails,
+    updateTriggeredDefences,
+  }: {
+    setEmails: (emails: any) => void;
```
Comment (gsproston-scottlogic) [inline frontend/src/components/ChatBox/ChatBox.tsx:18]: Could we type emails here with `OpenAIEmail`?

File: frontend/src/components/DefenceBox/DefenceBox.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+function DefenceBox({ triggeredDefences }: { triggeredDefences: string[] }) {
+  // list of defence mechanisms
+  const [defenceDetails, setDefenceDetails] = useState(
+    DEFENCE_DETAILS.map((defence) => {
+      return {
+        ...defence,
+        isActive: false,
+        isTriggered: false,
+      };
+    })
```
Comment (gsproston-scottlogic) [inline frontend/src/components/DefenceBox/DefenceBox.tsx:21]: If we have `isActive` and `isTriggered` default to `false` in the interface definition, we don't need this `map` here.

**Review by gsproston-scottlogic — CHANGES_REQUESTED** [review]
Looks good! I do agree that we should change any references to 'OpenAI' to something else. I reckon if it's related to the LLM, then change it to `chatBot...`, and just remove it from the email and defence interfaces.

File: frontend/src/service/defenceService.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import { sendRequest } from "./backendService";
+
+const PATH = "defence/";
+
+interface OpenAIDefence {
+  name: string;
+  type: string;
+  info: string;
+  isActive?: boolean;
+  isTriggered?: boolean;
```
[the remark this answers is not in the document]
Comment (gsproston-scottlogic) [reply frontend/src/service/defenceService.ts]: Ended up using a class with a constructor here as it doesn't look like it's possible to have default values in a TS interface.

**Review by gsproston-scottlogic — APPROVED** [review]

**Review by heatherlogan-scottlogic — APPROVED** [review]

**Pull request merged**